File: 0269-alien-dictionary/0269-alien-dictionary.py
class Solution:
    def alienOrder(self, w: List[str]) -> str:
        g = defaultdict(list)

        distinct = set()
        indeg = defaultdict(int)

        for x in w:
            for c in x:
                distinct.add(c)
                indeg[c] = 0
        if len(distinct) == 1: return next(iter(distinct))
        prev = w[0]
        n = len(w)

        for curr in w[1:]:
            i = 0
            while i < min(len(curr), len(prev)):
                if prev[i] == curr[i]:
                    i += 1
                else:
                    break
            if i >= len(curr) and i < len(prev):
                return ""
            if i  < len(curr) and i < len(prev):
                g[prev[i]].append(curr[i])
                indeg[curr[i]] += 1
                # if prev[i] not in indeg:
                #     indeg[prev[i]] = 0
            prev = curr
            
        q = deque([])
        vis = set()
        # print(indeg, g)
        for k, v in indeg.items():
            if v == 0:
                q.append(k)
                vis.add(k)

        stk = []

        while q:
            node = q.popleft()
            stk.append(node)
            for nei in g[node]:
                indeg[nei] -= 1
                if not indeg[nei] and nei not in vis:
                    q.append(nei)
                    vis.add(nei)

        # print(distinct, stk)
        if len(stk) != len(distinct):
            return ""


        return "".join(stk)
                


        


            
        

File: 0269-alien-dictionary/0269-alien-dictionary.java
class Solution {
    // DFS Helper (unchanged logic, just cleaner parameters)
    boolean dfs(Character node, Map<Character, List<Character>> g, Stack<Character> s, Set<Character> vis, Set<Character> path) {
        vis.add(node);
        path.add(node);
        
        for (Character nei : g.get(node)) { // g.get(node) is safe now because we initialize everything
            if (path.contains(nei)) return true; // Cycle detected
            if (!vis.contains(nei)) {
                if (dfs(nei, g, s, vis, path)) return true;
            }
        }

        s.push(node);
        path.remove(node);
        return false;
    }

    public String alienOrder(String[] words) {
        Map<Character, List<Character>> g = new HashMap<>();

        // 1. Initialize the graph with EVERY unique character (Matches Python logic)
        // This handles cases like ["z", "z"] where no edges are ever created.
        for (String word : words) {
            for (char c : word.toCharArray()) {
                g.putIfAbsent(c, new ArrayList<>());
            }
        }

        // 2. Build the edges
        for (int i = 0; i < words.length - 1; i++) {
            String w1 = words[i];
            String w2 = words[i + 1];
            
            // Check for invalid prefix case (e.g., "abc" before "ab")
            if (w1.length() > w2.length() && w1.startsWith(w2)) {
                return "";
            }

            // Find the first difference
            for (int j = 0; j < Math.min(w1.length(), w2.length()); j++) {
                if (w1.charAt(j) != w2.charAt(j)) {
                    g.get(w1.charAt(j)).add(w2.charAt(j));
                    break; // Only the first difference matters
                }
            }
            // Note: We removed the "if (!foundDiff) return """ block. 
            // If words are identical (["z", "z"]), we simply add no edges and continue.
        }

        Stack<Character> s = new Stack<>();
        Set<Character> vis = new HashSet<>();
        Set<Character> path = new HashSet<>();

        // 3. DFS over every character in the map
        for (Character c : g.keySet()) {
            if (!vis.contains(c)) {
                if (dfs(c, g, s, vis, path)) {
                    return ""; // Cycle detected
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        while (!s.isEmpty()) {
            sb.append(s.pop());
        }
        return sb.toString();
    }
}